File: src/com/Torvald/Terrarum/MapDrawer/MapCamera.java
package com.Torvald.Terrarum.MapDrawer;

import com.Torvald.Terrarum.Actors.Player;
import com.Torvald.Terrarum.Terrarum;
import com.Torvald.Terrarum.Game;
import com.Torvald.Terrarum.GameMap.GameMap;
import com.Torvald.Terrarum.GameMap.MapLayer;
import com.jme3.math.FastMath;
import org.newdawn.slick.*;

/**
 * Created by minjaesong on 16-01-19.
 */
public class MapCamera {

    private static GameMap map;
    private static Player player;

    private static int cameraX = 0;
    private static int cameraY = 0;

    private static SpriteSheet tilesWall;
    private static SpriteSheet tilesTerrain;
    private static SpriteSheet tilesWire;

    private static int TSIZE = MapDrawer.TILE_SIZE;


    private static SpriteSheet[] tilesetBook;

    private static final int WALL = 0;
    private static final int TERRAIN = 1;
    private static final int WIRE = 2;

    private static int renderWidth;
    private static int renderHeight;

    private static final int TILE_AIR = 0;

    private static final int NEARBY_TILE_KEY_UP = 0;
    private static final int NEARBY_TILE_KEY_RIGHT = 1;
    private static final int NEARBY_TILE_KEY_DOWN = 2;
    private static final int NEARBY_TILE_KEY_LEFT = 3;

    private static final int NEARBY_TILE_CODE_UP = 0b0001;
    private static final int NEARBY_TILE_CODE_RIGHT = 0b0010;
    private static final int NEARBY_TILE_CODE_DOWN = 0b0100;
    private static final int NEARBY_TILE_CODE_LEFT = 0b1000;

    /**
     * @param map
     * @param tileSize
     */
    public MapCamera(GameMap map, int tileSize) throws SlickException {
        this.map = map;
        player = Game.getPlayer();

        tilesWall = new SpriteSheet("./res/graphics/terrain/wall.png"
                , TSIZE
                , TSIZE
        );

        tilesTerrain = new SpriteSheet("./res/graphics/terrain/terrainplusplus.png"
                , TSIZE
                , TSIZE
        );

        tilesWire = new SpriteSheet("./res/graphics/terrain/wire.png"
                , TSIZE
                , TSIZE
        );

        tilesetBook = new SpriteSheet[9];
        tilesetBook[WALL] = tilesWall;
        tilesetBook[TERRAIN] = tilesTerrain;
        tilesetBook[WIRE] = tilesWire;
    }

    public static void update(GameContainer gc, int delta_t) {
        renderWidth = FastMath.ceil(Terrarum.WIDTH / Game.screenZoom);
        renderHeight = FastMath.ceil(Terrarum.HEIGHT / Game.screenZoom);

        // position - (WH / 2)
        cameraX = clamp(
                Math.round(player.pointedPosX() - (renderWidth / 2))
                , map.width * TSIZE - renderWidth
        );
        cameraY = clamp(
                Math.round(player.pointedPosY() - (renderHeight / 2))
                , map.height * TSIZE - renderHeight
        );
    }

    public static void render(GameContainer gc, Graphics g) {
        /**
         * render to camera
         */
        drawTiles(WALL);
        drawTiles(TERRAIN);
    }

    private static void drawTiles(int mode) {
        int for_y_start = div16(cameraY);
        int for_x_start = div16(cameraX);

        int for_y_end = clampHTile(for_y_start + div16(renderHeight) + 2);
        int for_x_end = clampWTile(for_x_start + div16(renderWidth) + 2);

        MapLayer currentLayer = (mode % 3 == WALL) ? map.getLayerWall()
                                               : (mode % 3 == TERRAIN) ? map.getLayerTerrain()
                                                                   : map.getLayerWire();

        // initialise
        tilesetBook[mode].startUse();

        // loop
        for (int y = for_y_start; y < for_y_end; y++) {
            for (int x = for_x_start; x < for_x_end; x++) {

                int thisTile = currentLayer.getTile(x, y);
                int thisTerrainTile = map.getTileFromTerrain(x, y);

                // draw
                if (

                        (
                                (       // wall and not blocked
                                        (mode == WALL) && (!isOpaque(thisTerrainTile))
                                )
                                        ||
                                        (mode == TERRAIN)
                        )       // not an air tile
                                && (thisTile > 0)
                                &&
                                // check if light level of upper tile is zero and
                                // that of this tile is also zero
                                (((y > 0)
                                        && !((LightmapRenderer.getValueFromMap(x, y) == 0)
                                                && (LightmapRenderer.getValueFromMap(x, y - 1) == 0))
                                )
                                ||
                                        // check if light level of this tile is zero, for y = 0
                                ((y == 0)
                                        && (LightmapRenderer.getValueFromMap(x, y) > 0)
                                        ))
                        ) {

                    if (mode == TERRAIN) {
                        int nearbyTilesInfo = getNearbyTilesInfo(x, y, TILE_AIR);

                        int thisTileX = nearbyTilesInfo;
                        int thisTileY = thisTile;

                        drawTile(TERRAIN, x, y, thisTileX, thisTileY);
                    }
                    else {
                        drawTile(mode, x, y, mod16(thisTile), div16(thisTile));
                    }
                }

            }
        }

        tilesetBook[mode].endUse();
    }

    /**
     *
     * @param x
     * @param y
     * @return [0-15] 1: up, 2: right, 4: down, 8: left
     */
    private static int getNearbyTilesInfo(int x, int y, int mark) {
        int[] nearbyTiles = new int[4];
        if (x == 0) { nearbyTiles[NEARBY_TILE_KEY_LEFT] = 0; }
        if (x == map.width - 1) { nearbyTiles[NEARBY_TILE_KEY_RIGHT] = 0; }
        if (y == 0) { nearbyTiles[NEARBY_TILE_KEY_UP] = 0; }
        if (y == map.height - 1) { nearbyTiles[NEARBY_TILE_KEY_DOWN] = 0; }
        try {
            nearbyTiles[NEARBY_TILE_KEY_UP] = map.getTileFromTerrain(x, y - 1);
            nearbyTiles[NEARBY_TILE_KEY_DOWN] = map.getTileFromTerrain(x, y + 1);
            nearbyTiles[NEARBY_TILE_KEY_LEFT] = map.getTileFromTerrain(x - 1, y);
            nearbyTiles[NEARBY_TILE_KEY_RIGHT] = map.getTileFromTerrain(x + 1, y);
        }
        catch (ArrayIndexOutOfBoundsException e) { }

        int ret = 0;
        for (int i = 0; i < 4; i++) {
            if (nearbyTiles[i] == mark) {
                ret += (1 << i); // write 1, 2, 4, 8 for i = 0, 1, 2, 3
            }
        }

        return ret;

    }

    private static void drawTile(int mode, int tilewisePosX, int tilewisePosY, int sheetX, int sheetY) {
        if (Game.screenZoom == 1) {
            tilesetBook[mode].renderInUse(
                    FastMath.floor(tilewisePosX * TSIZE)
                    , FastMath.floor(tilewisePosY * TSIZE)
                    , sheetX
                    , sheetY
            );
        }
        else {
            tilesetBook[mode].getSprite(
                    sheetX
                    , sheetY
            ).drawEmbedded(
                    Math.round(tilewisePosX * TSIZE * Game.screenZoom)
                    , Math.round(tilewisePosY * TSIZE * Game.screenZoom)
                    , FastMath.ceil(TSIZE * Game.screenZoom)
                    , FastMath.ceil(TSIZE * Game.screenZoom)
            );
        }
    }

    private static int div16(int x) {
        return (x & 0x7FFF_FFFF) >> 4;
    }

    private static int mod16(int x) {
        return x & 0b1111;
    }

    private static int quantise16(int x) {
        return (x & 0xFFFF_FFF0);
    }

    private static int clampW(int x) {
        if (x < 0) {
            return 0;
        }
        else if (x > map.width * TSIZE) {
            return map.width * TSIZE;
        }
        else {
            return x;
        }
    }

    private static int clampH(int x) {
        if (x < 0) {
            return 0;
        }
        else if (x > map.height * TSIZE) {
            return map.height * TSIZE;
        }
        else {
            return x;
        }
    }

    private static int clampWTile(int x) {
        if (x < 0) {
            return 0;
        }
        else if (x > map.width) {
            return map.width;
        }
        else {
            return x;
        }
    }

    private static int clampHTile(int x) {
        if (x < 0) {
            return 0;
        }
        else if (x > map.height) {
            return map.height;
        }
        else {
            return x;
        }
    }

    private static int clamp(int x, int lim) {
        if (x < 0) {
            return 0;
        }
        else if (x > lim) {
            return lim;
        }
        else {
            return x;
        }
    }

    private static Image getTileByIndex(SpriteSheet s, int i) {
        return s.getSprite(i % 16, i / 16);
    }

    private static boolean isOpaque(int x) {
        return (x >= 1 && x <= 38)
                || (x >= 41 && x <= 44)
                || (x >= 46 && x <= 47)
                || (x >= 64 && x <= 86)
                || (x >= 88 && x <= 116);
    }

    public static int getCameraX() {
        return cameraX;
    }

    public static int getCameraY() {
        return cameraY;
    }

    public static int getRenderWidth() {
        return renderWidth;
    }

    public static int getRenderHeight() {
        return renderHeight;
    }

    public static int getRenderStartX() {
        return div16(cameraX);
    }

    public static int getRenderStartY() {
        return div16(cameraY);
    }

    public static int getRenderEndX() {
        return clampWTile(getRenderStartX() + div16(renderWidth) + 2);
    }

    public static int getRenderEndY() {
        return clampHTile(getRenderStartY() + div16(renderHeight) + 2);
    }
}

File: src/com/Torvald/Terrarum/MapDrawer/MapCamera.kt
package com.torvald.terrarum.mapdrawer

import com.torvald.terrarum.gamemap.GameMap
import com.torvald.terrarum.gamemap.PairedMapLayer
import com.torvald.terrarum.Terrarum
import com.torvald.terrarum.tileproperties.TileNameCode
import com.torvald.terrarum.tileproperties.TilePropCodex
import com.jme3.math.FastMath
import com.torvald.terrarum.setBlendMul
import com.torvald.terrarum.setBlendNormal
import org.lwjgl.opengl.GL11
import org.newdawn.slick.GameContainer
import org.newdawn.slick.Graphics
import org.newdawn.slick.SlickException
import org.newdawn.slick.SpriteSheet
import java.util.*

/**
 * Created by minjaesong on 16-01-19.
 */
object MapCamera {
    private val map: GameMap = Terrarum.game.map;

    var cameraX = 0
        private set
    var cameraY = 0
        private set

    private val TSIZE = MapDrawer.TILE_SIZE

    private var tilesWall: SpriteSheet = SpriteSheet("./res/graphics/terrain/wall.png", TSIZE, TSIZE)
    private var tilesTerrain: SpriteSheet = SpriteSheet("./res/graphics/terrain/terrain.png", TSIZE, TSIZE)
    private var tilesWire: SpriteSheet = SpriteSheet("./res/graphics/terrain/wire.png", TSIZE, TSIZE)
    private var tilesetBook: Array<SpriteSheet> = arrayOf(tilesWall, tilesTerrain, tilesWire)

    private val WALL = GameMap.WALL
    private val TERRAIN = GameMap.TERRAIN
    private val WIRE = GameMap.WIRE

    private var renderWidth: Int = 0
    private var renderHeight: Int = 0

    private val NEARBY_TILE_KEY_UP = 0
    private val NEARBY_TILE_KEY_RIGHT = 1
    private val NEARBY_TILE_KEY_DOWN = 2
    private val NEARBY_TILE_KEY_LEFT = 3

    private val NEARBY_TILE_CODE_UP = 1
    private val NEARBY_TILE_CODE_RIGHT = 2
    private val NEARBY_TILE_CODE_DOWN = 4
    private val NEARBY_TILE_CODE_LEFT = 8

    /**
     * Connectivity group 01 : artificial tiles
     * It holds different shading rule to discriminate with group 02, index 0 is single tile.
     * These are the tiles that only connects to itself, will not connect to colour variants
     */
    private val TILES_CONNECT_SELF = arrayOf(
              TileNameCode.ICE_MAGICAL
            , TileNameCode.ILLUMINATOR_BLACK
            , TileNameCode.ILLUMINATOR_BLUE
            , TileNameCode.ILLUMINATOR_BROWN
            , TileNameCode.ILLUMINATOR_CYAN
            , TileNameCode.ILLUMINATOR_FUCHSIA
            , TileNameCode.ILLUMINATOR_GREEN
            , TileNameCode.ILLUMINATOR_GREEN_DARK
            , TileNameCode.ILLUMINATOR_GREY_DARK
            , TileNameCode.ILLUMINATOR_GREY_LIGHT
            , TileNameCode.ILLUMINATOR_GREY_MED
            , TileNameCode.ILLUMINATOR_ORANGE
            , TileNameCode.ILLUMINATOR_PURPLE
            , TileNameCode.ILLUMINATOR_RED
            , TileNameCode.ILLUMINATOR_TAN
            , TileNameCode.ILLUMINATOR_WHITE
            , TileNameCode.ILLUMINATOR_YELLOW
            , TileNameCode.ILLUMINATOR_BLACK_OFF
            , TileNameCode.ILLUMINATOR_BLUE_OFF
            , TileNameCode.ILLUMINATOR_BROWN_OFF
            , TileNameCode.ILLUMINATOR_CYAN_OFF
            , TileNameCode.ILLUMINATOR_FUCHSIA_OFF
            , TileNameCode.ILLUMINATOR_GREEN_OFF
            , TileNameCode.ILLUMINATOR_GREEN_DARK_OFF
            , TileNameCode.ILLUMINATOR_GREY_DARK_OFF
            , TileNameCode.ILLUMINATOR_GREY_LIGHT_OFF
            , TileNameCode.ILLUMINATOR_GREY_MED_OFF
            , TileNameCode.ILLUMINATOR_ORANGE_OFF
            , TileNameCode.ILLUMINATOR_PURPLE_OFF
            , TileNameCode.ILLUMINATOR_RED_OFF
            , TileNameCode.ILLUMINATOR_TAN_OFF
            , TileNameCode.ILLUMINATOR_WHITE_OFF
            , TileNameCode.ILLUMINATOR_YELLOW
            , TileNameCode.SANDSTONE
            , TileNameCode.SANDSTONE_BLACK
            , TileNameCode.SANDSTONE_DESERT
            , TileNameCode.SANDSTONE_RED
            , TileNameCode.SANDSTONE_WHITE
            , TileNameCode.SANDSTONE_GREEN
    )

    /**
     * Connectivity group 02 : natural tiles
     * It holds different shading rule to discriminate with group 01, index 0 is middle tile.
     */
    private val TILES_CONNECT_MUTUAL = arrayOf(
              TileNameCode.STONE
            , TileNameCode.DIRT
            , TileNameCode.GRASS
            , TileNameCode.PLANK_BIRCH
            , TileNameCode.PLANK_BLOODROSE
            , TileNameCode.PLANK_EBONY
            , TileNameCode.PLANK_NORMAL
            , TileNameCode.SAND
            , TileNameCode.SAND_WHITE
            , TileNameCode.SAND_RED
            , TileNameCode.SAND_DESERT
            , TileNameCode.SAND_BLACK
            , TileNameCode.SAND_GREEN
            , TileNameCode.GRAVEL
            , TileNameCode.GRAVEL_GREY
            , TileNameCode.SNOW
            , TileNameCode.ICE_NATURAL
            , TileNameCode.ORE_COPPER
            , TileNameCode.ORE_IRON
            , TileNameCode.ORE_GOLD
            , TileNameCode.ORE_SILVER
            , TileNameCode.ORE_ILMENITE
            , TileNameCode.ORE_AURICHALCUM

            , TileNameCode.WATER
            , TileNameCode.WATER_1
            , TileNameCode.WATER_2
            , TileNameCode.WATER_3
            , TileNameCode.WATER_4
            , TileNameCode.WATER_5
            , TileNameCode.WATER_6
            , TileNameCode.WATER_7
            , TileNameCode.WATER_8
            , TileNameCode.WATER_9
            , TileNameCode.WATER_10
            , TileNameCode.WATER_11
            , TileNameCode.WATER_12
            , TileNameCode.WATER_13
            , TileNameCode.WATER_14
            , TileNameCode.WATER_15
            , TileNameCode.LAVA
            , TileNameCode.LAVA_1
            , TileNameCode.LAVA_2
            , TileNameCode.LAVA_3
            , TileNameCode.LAVA_4
            , TileNameCode.LAVA_5
            , TileNameCode.LAVA_6
            , TileNameCode.LAVA_7
            , TileNameCode.LAVA_8
            , TileNameCode.LAVA_9
            , TileNameCode.LAVA_10
            , TileNameCode.LAVA_11
            , TileNameCode.LAVA_12
            , TileNameCode.LAVA_13
            , TileNameCode.LAVA_14
            , TileNameCode.LAVA_15
    )

    /**
     * Torches, levers, switches, ...
     */
    private val TILES_WALL_STICKER = arrayOf(
              TileNameCode.TORCH
            , TileNameCode.TORCH_FROST
            , TileNameCode.TORCH_OFF
            , TileNameCode.TORCH_FROST_OFF
    )

    /**
     * platforms, ...
     */
    private val TILES_WALL_STICKER_CONNECT_SELF = arrayOf<Int>(

    )

    /**
     * Tiles that half-transparent and has hue
     * will blend colour using colour multiplication
     * i.e. red hues get lost if you dive into the water
     */
    private val TILES_BLEND_MUL = arrayOf(
              TileNameCode.WATER
            , TileNameCode.WATER_1
            , TileNameCode.WATER_2
            , TileNameCode.WATER_3
            , TileNameCode.WATER_4
            , TileNameCode.WATER_5
            , TileNameCode.WATER_6
            , TileNameCode.WATER_7
            , TileNameCode.WATER_8
            , TileNameCode.WATER_9
            , TileNameCode.WATER_10
            , TileNameCode.WATER_11
            , TileNameCode.WATER_12
            , TileNameCode.WATER_13
            , TileNameCode.WATER_14
            , TileNameCode.WATER_15
            , TileNameCode.LAVA
            , TileNameCode.LAVA_1
            , TileNameCode.LAVA_2
            , TileNameCode.LAVA_3
            , TileNameCode.LAVA_4
            , TileNameCode.LAVA_5
            , TileNameCode.LAVA_6
            , TileNameCode.LAVA_7
            , TileNameCode.LAVA_8
            , TileNameCode.LAVA_9
            , TileNameCode.LAVA_10
            , TileNameCode.LAVA_11
            , TileNameCode.LAVA_12
            , TileNameCode.LAVA_13
            , TileNameCode.LAVA_14
            , TileNameCode.LAVA_15
    )

    fun update(gc: GameContainer, delta_t: Int) {
        val player = Terrarum.game.player

        renderWidth = FastMath.ceil(Terrarum.WIDTH / Terrarum.game.screenZoom) // div, not mul
        renderHeight = FastMath.ceil(Terrarum.HEIGHT / Terrarum.game.screenZoom)

        // position - (WH / 2)
        cameraX = Math.round(FastMath.clamp(
                player.hitbox!!.centeredX - renderWidth / 2, TSIZE.toFloat(), map.width * TSIZE - renderWidth - TSIZE.toFloat()))
        cameraY = Math.round(FastMath.clamp(
                player.hitbox!!.centeredY - renderHeight / 2, TSIZE.toFloat(), map.height * TSIZE - renderHeight - TSIZE.toFloat()))
    }

    fun renderBehind(gc: GameContainer, g: Graphics) {
        /**
         * render to camera
         */
        setBlendNormal()
        drawTiles(WALL, false)
        drawTiles(TERRAIN, false)
    }

    fun renderFront(gc: GameContainer, g: Graphics) {
        setBlendMul()
        drawTiles(TERRAIN, true)
        setBlendNormal()
    }

    private fun drawTiles(mode: Int, drawModeTilesBlendMul: Boolean) {
        val for_y_start = div16(cameraY)
        val for_x_start = div16(cameraX)

        val for_y_end = clampHTile(for_y_start + div16(renderHeight) + 2)
        val for_x_end = clampWTile(for_x_start + div16(renderWidth) + 2)

        // initialise
        tilesetBook[mode].startUse()

        // loop
        for (y in for_y_start..for_y_end - 1) {
            for (x in for_x_start..for_x_end - 1) {

                val thisTile: Int?
                if (mode % 3 == WALL)
                    thisTile = map.getTileFromWall(x, y)
                else if (mode % 3 == TERRAIN)
                    thisTile = map.getTileFromTerrain(x, y)
                else if (mode % 3 == WIRE)
                    thisTile = map.getTileFromWire(x, y)
                else
                    throw IllegalArgumentException()

                val noDamageLayer = mode % 3 == WIRE

                // draw
                try {
                    if (

                    (mode == WALL || mode == TERRAIN)       // not an air tile

                            && (thisTile ?: 0) > 0
                            &&
                            // check if light level of nearby or this tile is illuminated
                    (    LightmapRenderer.getValueFromMap(x, y) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x - 1, y) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x + 1, y) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x, y - 1) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x, y + 1) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x - 1, y - 1) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x + 1, y + 1) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x + 1, y - 1) ?: 0 > 0
                         || LightmapRenderer.getValueFromMap(x - 1, y + 1) ?: 0 > 0)
                    ) {

                        val nearbyTilesInfo: Int
                        if (isWallSticker(thisTile)) {
                            nearbyTilesInfo = getNearbyTilesInfoWallSticker(x, y)
                        } else if (isConnectMutual(thisTile)) {
                            nearbyTilesInfo = getNearbyTilesInfoNonSolid(x, y, mode)
                        } else if (isConnectSelf(thisTile)) {
                            nearbyTilesInfo = getNearbyTilesInfo(x, y, mode, thisTile)
                        } else {
                            nearbyTilesInfo = 0
                        }


                        val thisTileX: Int
                        if (!noDamageLayer)
                            thisTileX = PairedMapLayer.RANGE * ((thisTile ?: 0) % PairedMapLayer.RANGE) + nearbyTilesInfo
                        else
                            thisTileX = nearbyTilesInfo

                        val thisTileY = (thisTile ?: 0) / PairedMapLayer.RANGE

                        if (drawModeTilesBlendMul) {
                            if (isBlendMul(thisTile)) {
                                drawTile(mode, x, y, thisTileX, thisTileY)
                            }
                        } else {
                            // do NOT add "if (!isBlendMul(thisTile))"!
                            // or else they will not look like they should be when backed with wall
                            drawTile(mode, x, y, thisTileX, thisTileY)
                        }
                    }
                } catch (e: NullPointerException) {
                    // do nothing. This exception handling may hide erratic behaviour completely.
                }

            }
        }

        tilesetBook[mode].endUse()
    }

    private fun getGrassInfo(x: Int, y: Int, from: Int, to: Int): Int {
        return 0
    }

    /**

     * @param x
     * *
     * @param y
     * *
     * @return [0-15] 1: up, 2: right, 4: down, 8: left
     */
    private fun getNearbyTilesInfo(x: Int, y: Int, mode: Int, mark: Int?): Int {
        val nearbyTiles = IntArray(4)
        nearbyTiles[NEARBY_TILE_KEY_LEFT] = map.getTileFrom(mode, x - 1, y) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_RIGHT] = map.getTileFrom(mode, x + 1, y) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_UP] = map.getTileFrom(mode, x, y - 1) ?: 4906
        nearbyTiles[NEARBY_TILE_KEY_DOWN] = map.getTileFrom(mode, x, y + 1) ?: 4096

        // try for
        var ret = 0
        for (i in 0..3) {
            if (nearbyTiles[i] == mark) {
                ret += 1 shl i // add 1, 2, 4, 8 for i = 0, 1, 2, 3
            }
        }

        return ret
    }

    private fun getNearbyTilesInfoNonSolid(x: Int, y: Int, mode: Int): Int {
        val nearbyTiles = IntArray(4)
        nearbyTiles[NEARBY_TILE_KEY_LEFT] = map.getTileFrom(mode, x - 1, y) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_RIGHT] = map.getTileFrom(mode, x + 1, y) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_UP] = map.getTileFrom(mode, x, y - 1) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_DOWN] = map.getTileFrom(mode, x, y + 1) ?: 4096

        // try for
        var ret = 0
        for (i in 0..3) {
            try {
                if (!TilePropCodex.getProp(nearbyTiles[i]).isSolid) {
                    ret += (1 shl i) // add 1, 2, 4, 8 for i = 0, 1, 2, 3
                }
            } catch (e: ArrayIndexOutOfBoundsException) {
            }

        }

        return ret
    }

    private fun getNearbyTilesInfoWallSticker(x: Int, y: Int): Int {
        val nearbyTiles = IntArray(4)
        val NEARBY_TILE_KEY_BACK = NEARBY_TILE_KEY_UP
        nearbyTiles[NEARBY_TILE_KEY_LEFT] = map.getTileFrom(TERRAIN, x - 1, y) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_RIGHT] = map.getTileFrom(TERRAIN, x + 1, y) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_DOWN] = map.getTileFrom(TERRAIN, x, y + 1) ?: 4096
        nearbyTiles[NEARBY_TILE_KEY_BACK] = map.getTileFrom(WALL, x, y) ?: 4096

        try {
            if (TilePropCodex.getProp(nearbyTiles[NEARBY_TILE_KEY_RIGHT]).isSolid
                    && TilePropCodex.getProp(nearbyTiles[NEARBY_TILE_KEY_LEFT]).isSolid) {
                if (TilePropCodex.getProp(nearbyTiles[NEARBY_TILE_KEY_BACK]).isSolid)
                    return 0
                else
                    return 3
            } else if (TilePropCodex.getProp(nearbyTiles[NEARBY_TILE_KEY_RIGHT]).isSolid) {
                return 2
            } else if (TilePropCodex.getProp(nearbyTiles[NEARBY_TILE_KEY_LEFT]).isSolid) {
                return 1
            } else if (TilePropCodex.getProp(nearbyTiles[NEARBY_TILE_KEY_BACK]).isSolid) {
                return 0
            } else
                return 3
        } catch (e: ArrayIndexOutOfBoundsException) {
            return if (TilePropCodex.getProp(nearbyTiles[NEARBY_TILE_KEY_BACK]).isSolid)
                0
            else
                3
        }

    }

    private fun drawTile(mode: Int, tilewisePosX: Int, tilewisePosY: Int, sheetX: Int, sheetY: Int) {
        if (Terrarum.game.screenZoom == 1f) {
            tilesetBook[mode].renderInUse(
                    FastMath.floor((tilewisePosX * TSIZE).toFloat()), FastMath.floor((tilewisePosY * TSIZE).toFloat()), sheetX, sheetY)
        } else {
            tilesetBook[mode].getSprite(
                    sheetX, sheetY).drawEmbedded(
                    Math.round(tilewisePosX.toFloat() * TSIZE.toFloat() * Terrarum.game.screenZoom).toFloat(), Math.round(tilewisePosY.toFloat() * TSIZE.toFloat() * Terrarum.game.screenZoom).toFloat(), FastMath.ceil(TSIZE * Terrarum.game.screenZoom).toFloat(), FastMath.ceil(TSIZE * Terrarum.game.screenZoom).toFloat())
        }
    }

    fun div16(x: Int): Int = x and 0x7FFFFFFF shr 4
    fun mod16(x: Int): Int = x and 15
    fun quantise16(x: Int): Int = x and 0xFFFFFFF0.toInt()

    fun clampW(x: Int): Int {
        if (x < 0) {
            return 0
        } else if (x > map.width * TSIZE) {
            return map.width * TSIZE
        } else {
            return x
        }
    }

    fun clampH(x: Int): Int {
        if (x < 0) {
            return 0
        } else if (x > map.height * TSIZE) {
            return map.height * TSIZE
        } else {
            return x
        }
    }

    fun clampWTile(x: Int): Int {
        if (x < 0) {
            return 0
        } else if (x > map.width) {
            return map.width
        } else {
            return x
        }
    }

    fun clampHTile(x: Int): Int {
        if (x < 0) {
            return 0
        } else if (x > map.height) {
            return map.height
        } else {
            return x
        }
    }

    fun getRenderWidth(): Int = renderWidth
    fun getRenderHeight(): Int = renderHeight

    fun getRenderStartX(): Int = div16(cameraX)
    fun getRenderStartY(): Int = div16(cameraY)

    fun getRenderEndX(): Int = clampWTile(getRenderStartX() + div16(renderWidth) + 2)
    fun getRenderEndY(): Int = clampHTile(getRenderStartY() + div16(renderHeight) + 2)

    private fun isConnectSelf(b: Int?): Boolean = TILES_CONNECT_SELF.contains(b)
    private fun isConnectMutual(b: Int?): Boolean = TILES_CONNECT_MUTUAL.contains(b)
    private fun isWallSticker(b: Int?): Boolean = TILES_WALL_STICKER.contains(b)
    private fun isPlatform(b: Int?): Boolean = TILES_WALL_STICKER_CONNECT_SELF.contains(b)

    private fun isBlendMul(b: Int?): Boolean = TILES_BLEND_MUL.contains(b)

}